File: app/src/main/java/ch/unstable/ost/api/model/DepartureCheckpoint.java
package ch.unstable.ost.api.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;

import java.util.Date;

import ch.unstable.ost.utils.ParcelUtils;

public class DepartureCheckpoint extends Checkpoint implements Parcelable {
    public static final Creator<DepartureCheckpoint> CREATOR = new Creator<DepartureCheckpoint>() {
        @Override
        public DepartureCheckpoint createFromParcel(Parcel in) {
            return new DepartureCheckpoint(in);
        }

        @Override
        public DepartureCheckpoint[] newArray(int size) {
            return new DepartureCheckpoint[size];
        }
    };
    private final Date departureTime;

    public DepartureCheckpoint(Date departureTime, String platform, Location location) {
        super(platform, location);
        this.departureTime = departureTime;
    }

    protected DepartureCheckpoint(Parcel in) {
        super(in);
        departureTime = ParcelUtils.readDate(in);
    }

    @NonNull
    @Override
    public Date getDisplayDate() {
        return departureTime;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        ParcelUtils.writeDate(dest, departureTime);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public Date getDepartureTime() {
        return departureTime;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        DepartureCheckpoint that = (DepartureCheckpoint) o;
        return Objects.equal(departureTime, that.departureTime);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(super.hashCode(), departureTime);
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("departureTime", departureTime)
                .add("location", getLocation())
                .add("platform", getPlatform())
                .toString();
    }
}

File: app/src/main/java/ch/unstable/ost/api/model/DepartureCheckpoint.kt
package ch.unstable.ost.api.model

import android.os.Parcel
import android.os.Parcelable
import ch.unstable.ost.utils.ParcelUtils
import java.util.*

data class DepartureCheckpoint(val departureTime: Date, val stopLocation: StopLocation) : Parcelable {

    constructor(departureTime: Date, platform: String?, location: Location):
            this(departureTime, StopLocation(platform, location));

    private constructor(source: Parcel): this(
            departureTime = ParcelUtils.readDate(source)!!,
            stopLocation = ParcelUtils.readParcelable(source, StopLocation.CREATOR)!!)

    override fun writeToParcel(dest: Parcel, flags: Int) {
        ParcelUtils.writeDate(dest, departureTime)
        ParcelUtils.writeParcelable(dest, stopLocation, flags)
    }

    override fun describeContents(): Int {
        return 0
    }


    companion object CREATOR: Parcelable.Creator<DepartureCheckpoint> {
        override fun createFromParcel(`in`: Parcel): DepartureCheckpoint {
            return DepartureCheckpoint(`in`)
        }

        override fun newArray(size: Int): Array<DepartureCheckpoint?> {
            return arrayOfNulls(size)
        }
    }

    val stationName: String
        get() = stopLocation.location.name

    val location: Location
        get() = stopLocation.location

    val platform: String?
        get() = stopLocation.platform
}
